xapian-core/api/registry.cc

```diff
@@ -107,6 +111,49 @@ register_object(map<string, T*> & registry, const T & obj)
 }
 
 template<class T>
+static inline void
+register_object(map<string, T*> & registry1, map<string, T*> & registry2,
+		const T & obj)
+{
+    string name = obj.name();
+    if (rare(name.empty())) {
+	throw Xapian::InvalidOperationError("Unable to register object - name() method returned empty string");
+    }
+
+    string short_name = obj.name();
+    if (rare(short_name.empty())) {
+	throw Xapian::InvalidOperationError("Unable to register object - short_name() method returned empty string");
+    }
+
+    pair<typename map<string, T *>::iterator, bool> r1;
+    r1 = registry1.insert(make_pair(name, static_cast<T*>(NULL)));
+    pair<typename map<string, T *>::iterator, bool> r2;
+    r2 = registry2.insert(make_pair(short_name, static_cast<T*>(NULL)));
+
+    if (!r1.second) {
+	// Existing element with this key, so replace the pointer with NULL
+	// and delete the existing pointer.
+	//
+	// If the delete throws, this will leave a NULL entry in the map, but
+	// that won't affect behaviour as we return NULL for "not found"
+	// anyway.  The memory used will be leaked if the dtor throws, but
+	// throwing exceptions from the dtor is bad form, so that's not a big
+	// problem.
+	T * p = NULL;
+	swap(p, r1.first->second);
+	delete p;
+    }
+
+    T * clone = obj.clone();
+    if (rare(!clone)) {
+	throw Xapian::InvalidOperationError("Unable to register object - clone() method returned NULL");
+    }
+
+    r1.first->second = clone;
+    r2.first->second = clone;
+}
```

Comment:
We're mostly there, but you don't seem to have addressed the wrinkle I mentioned here with sharing the object between the two maps.

To re-iterate, if I register a `Weight` object which reports the same `name()` as an already registered object but a different `short_name()`, then the object in the name map is replaced (and so the old object deleted) but the entry in the short name map for the old short name is left, and that's now a dangling pointer.  This is an obscure case, but we really shouldn't leave dangling pointers like this where they could get used as they're a potential security risk.

We need to take a bit more care to deal with this situation.

One option (the one I mentioned before) is to reject attempts to do this (and throw `Xapian::InvalidOperatorError` instead), which has the advantage of not letting people create a situation like this accidentally.  To do that we need to check that (a) entries for the two names either both do or both don't already exist (because if only one of them exists, a previous object used the same short or long name, but not both, and (b) if they both do both exist, that the pointers in them are the same (because otherwise we're trying to replace the long name entry for one weight and the short name entry for a different one).  I think the correct check is something like: `(r1.second != r2.second || (r1.second && r1.first->second != r2.first->second))`

We could reliably remove the old object from the short name map instead - in the case where we replacing the object in the name map, before we delete the old object we need to:

* check if we're replacing an existing object in the short_name map too, and if so if that object is the same one we're replacing in the name map - if so, we can just continue as you already do.
* otherwise, we need to locate the corresponding short_name entry and remove it from the short_name map (at least if there is such an entry - it could be it got overwritten by registering a second weighting scheme with the same short name) - we can do this by doing a lookup using the short name from the object we're about to delete, and then comparing pointers to make sure it is the same object (another could have been registered over the top).

So that would be something like:

```
    if (!r1.second) {
	// ...
	T * p = NULL;
	swap(p, r1.first->second);
        if (p && (r2.second || r2.first->second != p)) {
            // We're replacing an object with a different short name to us,
            // so check if the old object is still registered under its short
            // name.
            auto it = registry2.find(p->short_name());
            if (it != registry2.end() && *it == p)
                registry2.erase(it);
        }
        delete p;
    }
```

But that's more complex, and just seems to enable creating a confusing situation.  Being able to updated a registered weighting scheme is probably useful, but being able to create a situation in which two registered objects have the same short name seems mostly a trap for the unwary.